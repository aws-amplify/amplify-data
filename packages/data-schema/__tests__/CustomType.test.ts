import { expectTypeTestsToPassAsync } from 'jest-tsd';
import { a } from '../index';

// evaluates type defs in corresponding test-d.ts file
it('should not produce static type errors', async () => {
  await expectTypeTestsToPassAsync(__filename);
});

describe('CustomType unsupported modifiers', () => {
  test('Runtime error when calling unsupported modifiers', () => {
    expect(() => {
      a.customType({
        lat: a.float(),
        long: a.float(),
        // @ts-expect-error
      }).required();
    }).toThrowError();

    expect(() => {
      a.customType({
        lat: a.float(),
        long: a.float(),
        // @ts-expect-error
      }).default();
    }).toThrowError();

    expect(() => {
      a.customType({
        lat: a.float(),
        long: a.float(),
        // @ts-expect-error
      }).array();
    }).toThrowError();
  });
});

describe('CustomType transform', () => {
  test('Explicit CustomType', () => {
    const s = a
      .schema({
        Post: a.model({
          location: a.ref('Location'),
        }),
        Location: a.customType({
          lat: a.float(),
          long: a.float(),
        }),
      })
      .authorization([a.allow.public()]);

    const result = s.transform().schema;

    expect(result).toMatchSnapshot();
  });

  test('Explicit CustomType - required', () => {
    const s = a
      .schema({
        Post: a.model({
          location: a.ref('Location').required(),
        }),
        Location: a.customType({
          lat: a.float(),
          long: a.float(),
        }),
      })
      .authorization([a.allow.public()]);

    const result = s.transform().schema;

    expect(result).toMatchSnapshot();
  });

  test('Explicit CustomType nests implicit enum type', () => {
    const s = a
      .schema({
        Post: a.model({
          meta: a.ref('Meta'),
        }),
        Meta: a.customType({
          status: a.enum(['unpublished', 'published']),
          publishedDate: a.date(),
        }),
      })
      .authorization([a.allow.public()]);

    const result = s.transform().schema;

    expect(result).toMatchSnapshot();
  });

  test('Explicit CustomType nests implicit CustomType', () => {
    const s = a
      .schema({
        Post: a.model({
          meta: a.ref('Meta'),
        }),
        Meta: a.customType({
          status: a.enum(['unpublished', 'published']),
          nestedMeta: a.customType({
            field1: a.string(),
          }),
        }),
      })
      .authorization([a.allow.public()]);

    const result = s.transform().schema;

    expect(result).toMatchSnapshot();
  });

  test('Explicit CustomType nests explicit enum', () => {
    const s = a
      .schema({
        Post: a.model({
          meta: a.ref('Meta'),
        }),
        Meta: a.customType({
          status: a.ref('PostStatus'),
          publishedDate: a.date(),
        }),
        PostStatus: a.enum(['unpublished', 'published']),
      })
      .authorization([a.allow.public()]);

    const result = s.transform().schema;

    expect(result).toMatchSnapshot();
  });

  test('Explicit CustomType nests explicit CustomType', () => {
    const s = a
      .schema({
        Post: a.model({
          meta: a.ref('Meta'),
        }),
        Meta: a.customType({
          status: a.ref('PostStatus'),
          nestedMeta: a.ref('AltMeta'),
        }),
        PostStatus: a.enum(['unpublished', 'published']),
        AltMeta: a.customType({
          field1: a.string(),
        }),
      })
      .authorization([a.allow.public()]);

    const result = s.transform().schema;

    expect(result).toMatchSnapshot();
  });

  test('Explicit CustomType - with auth', () => {
    const s = a
      .schema({
        Post: a.model({
          location: a.ref('Location').authorization([a.allow.owner()]),
        }),
        Location: a.customType({
          lat: a.float(),
          long: a.float(),
        }),
      })
      .authorization([a.allow.public()]);

    const result = s.transform().schema;

    expect(result).toMatchSnapshot();
  });

  test('Implicit CustomType', () => {
    const s = a
      .schema({
        Post: a.model({
          location: a.customType({
            lat: a.float(),
            long: a.float(),
          }),
        }),
      })
      .authorization([a.allow.public()]);

    const result = s.transform().schema;

    expect(result).toMatchSnapshot();
  });

  test('Implicit CustomType nests implicit enum type', () => {
    const s = a
      .schema({
        Post: a.model({
          meta: a.customType({
            status: a.enum(['unpublished', 'published']),
            publishedDate: a.date(),
          }),
        }),
      })
      .authorization([a.allow.public()]);

    const result = s.transform().schema;

    expect(result).toMatchSnapshot();
  });

  test('Implicit CustomType nests implicit CustomType', () => {
    const s = a
      .schema({
        Post: a.model({
          meta: a.customType({
            status: a.enum(['unpublished', 'published']),
            nestedMeta: a.customType({
              field1: a.string(),
            }),
          }),
        }),
      })
      .authorization([a.allow.public()]);

    const result = s.transform().schema;

    expect(result).toMatchSnapshot();
  });

  test('Implicit CustomType nests explicit enum', () => {
    const s = a
      .schema({
        Post: a.model({
          meta: a.customType({
            status: a.ref('PostStatus'),
            publishedDate: a.date(),
          }),
        }),
        PostStatus: a.enum(['unpublished', 'published']),
      })
      .authorization([a.allow.public()]);

    const result = s.transform().schema;

    expect(result).toMatchSnapshot();
  });

  test('Implicit CustomType nests explicit CustomType', () => {
    const s = a
      .schema({
        Post: a.model({
          meta: a.customType({
            status: a.enum(['unpublished', 'published']),
            nestedMeta: a.ref('AltMeta'),
          }),
        }),
        AltMeta: a.customType({
          field1: a.string(),
        }),
      })
      .authorization([a.allow.public()]);

    const result = s.transform().schema;

    expect(result).toMatchSnapshot();
  });
});
