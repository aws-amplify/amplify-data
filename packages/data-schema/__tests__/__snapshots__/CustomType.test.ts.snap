// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomType transform Explicit CustomType - required 1`] = `
"type Post @model @auth(rules: [{allow: public, provider: apiKey}])
{
  location: Location!
}

type Location 
{
  lat: Float
  long: Float
}"
`;

exports[`CustomType transform Explicit CustomType - with auth 1`] = `
"type Post @model @auth(rules: [{allow: public, provider: apiKey}])
{
  location: Location @auth(rules: [{allow: owner, ownerField: "owner"}])
}

type Location 
{
  lat: Float
  long: Float
}"
`;

exports[`CustomType transform Explicit CustomType 1`] = `
"type Post @model @auth(rules: [{allow: public, provider: apiKey}])
{
  location: Location
}

type Location 
{
  lat: Float
  long: Float
}"
`;

exports[`CustomType transform Explicit CustomType nests explicit CustomType 1`] = `
"type Post @model @auth(rules: [{allow: public, provider: apiKey}])
{
  meta: Meta
}

type Meta 
{
  status: PostStatus
  nestedMeta: AltMeta
}

type AltMeta 
{
  field1: String
}

enum PostStatus {
  unpublished
  published
}"
`;

exports[`CustomType transform Explicit CustomType nests explicit enum 1`] = `
"type Post @model @auth(rules: [{allow: public, provider: apiKey}])
{
  meta: Meta
}

type Meta 
{
  status: PostStatus
  publishedDate: AWSDate
}

enum PostStatus {
  unpublished
  published
}"
`;

exports[`CustomType transform Explicit CustomType nests implicit CustomType 1`] = `
"type Post @model @auth(rules: [{allow: public, provider: apiKey}])
{
  meta: Meta
}

type Meta 
{
  status: MetaStatus
  nestedMeta: MetaNestedMeta
}

type MetaNestedMeta 
{
  field1: String
}

enum MetaStatus {
  unpublished
  published
}"
`;

exports[`CustomType transform Explicit CustomType nests implicit enum type 1`] = `
"type Post @model @auth(rules: [{allow: public, provider: apiKey}])
{
  meta: Meta
}

type Meta 
{
  status: MetaStatus
  publishedDate: AWSDate
}

enum MetaStatus {
  unpublished
  published
}"
`;

exports[`CustomType transform Implicit CustomType 1`] = `
"type Post @model @auth(rules: [{allow: public, provider: apiKey}])
{
  location: PostLocation
}

type PostLocation 
{
  lat: Float
  long: Float
}"
`;

exports[`CustomType transform Implicit CustomType nests explicit CustomType 1`] = `
"type Post @model @auth(rules: [{allow: public, provider: apiKey}])
{
  meta: PostMeta
}

type AltMeta 
{
  field1: String
}

type PostMeta 
{
  status: PostMetaStatus
  nestedMeta: AltMeta
}

enum PostMetaStatus {
  unpublished
  published
}"
`;

exports[`CustomType transform Implicit CustomType nests explicit enum 1`] = `
"type Post @model @auth(rules: [{allow: public, provider: apiKey}])
{
  meta: PostMeta
}

type PostMeta 
{
  status: PostStatus
  publishedDate: AWSDate
}

enum PostStatus {
  unpublished
  published
}"
`;

exports[`CustomType transform Implicit CustomType nests implicit CustomType 1`] = `
"type Post @model @auth(rules: [{allow: public, provider: apiKey}])
{
  meta: PostMeta
}

type PostMeta 
{
  status: PostMetaStatus
  nestedMeta: PostMetaNestedMeta
}

type PostMetaNestedMeta 
{
  field1: String
}

enum PostMetaStatus {
  unpublished
  published
}"
`;

exports[`CustomType transform Implicit CustomType nests implicit enum type 1`] = `
"type Post @model @auth(rules: [{allow: public, provider: apiKey}])
{
  meta: PostMeta
}

type PostMeta 
{
  status: PostMetaStatus
  publishedDate: AWSDate
}

enum PostMetaStatus {
  unpublished
  published
}"
`;
